bolt-sidecar/src/api/commitments/firewall_recv.rs

```diff
@@ -0,0 +1,622 @@
+            };
+
+        if !should_retry {
+            warn!("shutting down connection to {}", url);
+            break;
+        }
+
+        // retry_with_backoff(MAX_RETRIES, fut)
+
+        // Reconnect on failure
+        // TODO: add backoff
+        warn!(?url, "connection lost. reconnecting...");
+        tokio::time::sleep(Duration::from_secs(5)).await;
+    }
+}
+
+type PendingCommitmentResult = dyn Future<Output = Result<Result<SignedCommitment, CommitmentError>, oneshot::error::RecvError>>
+    + Send;
+
+/// The [CommitmentRequestProcessor] handles incoming commitment requests a the websocket
+/// connection, and forwards them to the [CommitmentEvent] tx channel for processing.
+struct CommitmentRequestProcessor {
+    /// The URL of the connected websocket server.
+    url: Url,
+    /// The channel to send commitment events to be processed.
+    api_events_tx: mpsc::Sender<CommitmentEvent>,
+    /// The websocket writer sink.
+    write_sink: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+    /// The websocket reader stream.
+    read_stream: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+    /// The receiver for keep-alive ping messages.
+    ping_rx: broadcast::Receiver<()>,
+    /// The receiver for shutdown signals.
+    shutdown_rx: broadcast::Receiver<()>,
+    /// The collection of pending commitment requests responses, sent with [api_events_tx].
+    /// NOTE: Is there a better way to avoid this monster type?
+    /// SAFETY: the `poll` implementation of this struct promptly handles these responses and
+    /// ensures this vector doesn't grow indefinitely.
+    pending_commitment_responses: FuturesUnordered<Pin<Box<PendingCommitmentResult>>>,
+    /// The collection of outgoing messages to be sent to the connected websocket server.
+    outgoing_messages: VecDeque<Message>,
+}
+
+impl CommitmentRequestProcessor {
+    pub fn new(
+        url: Url,
+        tx: mpsc::Sender<CommitmentEvent>,
+        write: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+        read: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+        ping_rx: broadcast::Receiver<()>,
+        shutdown_rx: broadcast::Receiver<()>,
+    ) -> Self {
+        Self {
+            url,
+            api_events_tx: tx,
+            write_sink: write,
+            read_stream: read,
+            ping_rx,
+            shutdown_rx,
+            pending_commitment_responses: FuturesUnordered::new(),
+            outgoing_messages: VecDeque::new(),
+        }
+    }
+}
+
+impl Future for CommitmentRequestProcessor {
+    // The output of this future is a boolean indicating whether reconnection is needed.
+    type Output = bool;
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        let rpc_url = this.url.clone();
+
+        loop {
+            let mut progress = false;
```

Comment:
This pattern polls everything concurrently, with the same priority (which is fine). What you could also do is first handle messages with high priority, such as incoming requests / outbound commitments, where you `continue` until all of those messages are processed. You can then poll lower priority stuff like pings and shutdowns after. Here probably not needed but wanted to bring it up.

Reply:
Reporting some of the internal discussion we had on the topic:
- there is already a priority expressed by continuously polling the `Stream` with `while let Poll::Ready(T)`.
- it's true that using `continue` allows for more granular logic and thanks for making me aware of it. I'll consider it in case of further optimizations of the polling logic in the future.